src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
             {
-                return AwaitMatch(this, httpContext, feature, matchTask);
+                return AwaitMatch(this, httpContext, feature, previousEndpoint, matchTask);
             }
 
-            return SetRoutingAndContinue(httpContext, feature);
+            return SetRoutingAndContinue(httpContext, feature, previousEndpoint);
 
             // Awaited fallbacks for when the Tasks do not synchronously complete
-            static async Task AwaitMatcher(EndpointRoutingMiddleware middleware, HttpContext httpContext, EndpointSelectorContext feature, Task<Matcher> matcherTask)
+            static async Task AwaitMatcher(EndpointRoutingMiddleware middleware,
+                                           HttpContext httpContext,
+                                           EndpointSelectorContext feature,
+                                           Endpoint previousEndpoint,
+                                           Task<Matcher> matcherTask)
             {
                 var matcher = await matcherTask;
                 await matcher.MatchAsync(httpContext, feature);
-                await middleware.SetRoutingAndContinue(httpContext, feature);
+                await middleware.SetRoutingAndContinue(httpContext, feature, previousEndpoint);
             }
 
-            static async Task AwaitMatch(EndpointRoutingMiddleware middleware, HttpContext httpContext, EndpointSelectorContext feature, Task matchTask)
+            static async Task AwaitMatch(EndpointRoutingMiddleware middleware,
+                                         HttpContext httpContext,
+                                         EndpointSelectorContext feature,
+                                         Endpoint previousEndpoint,
+                                         Task matchTask)
             {
                 await matchTask;
-                await middleware.SetRoutingAndContinue(httpContext, feature);
+                await middleware.SetRoutingAndContinue(httpContext, feature, previousEndpoint);
             }
 
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private Task SetRoutingAndContinue(HttpContext httpContext, EndpointSelectorContext feature)
+        private Task SetRoutingAndContinue(HttpContext httpContext, EndpointSelectorContext feature, Endpoint previousEndpoint)
         {
-            if (feature.Endpoint != null)
-            {
-                // Set the endpoint feature only on success. This means we won't overwrite any
-                // existing state for related features unless we did something.
-                SetFeatures(httpContext, feature);
+            // Restore the previous endpoint (which might be null) if it's not set
+            feature.Endpoint ??= previousEndpoint;
 
-                Log.MatchSuccess(_logger, feature);
+            // If there was no mutation of the endpoint then log failure
+            if (ReferenceEquals(feature.Endpoint, previousEndpoint))
```

Comment:
I think this is wrong. The previous endpoint could be the same as the match endpoint, e.g. someone could:
1. Create a RoutingEndpoint in `UseEndpoints`
2. In middleware prior to routing they could fetch it from the endpoint datasource and set it on the feature
3. Routing middleware matches the endpoint already on feature

feature.Endpoint is set to null at the start of routing. If it is null at the end then there was no match. Check for null here and then restore the previousEndpoint inside the if.

Reply:
Add a unit test for this

Reply:
Sounds like an extreme edge case but so that I’m clear, pick the exact same instance that would have matched the endpoint and observe that the logs are bad yes?